nwb_conversion_tools/utils/basenwbephyswriter.py

```diff
@@ -0,0 +1,687 @@
+                        self._conversion_ops["skip_unit_features"].append(ft)
+                        print(f"Skipping feature '{ft}' because it is not defined for all spikes.")
+                        break
+                        # this means features are available for a subset of spikes
+                        # all_feat_vals = np.array([np.nan] * nspikes[unit_id])
+                        # feature_idxs = sorting.get_unit_spike_features(unit_id, feat_name + '_idxs')
+                        # all_feat_vals[feature_idxs] = feat_vals
+                    else:
+                        values.append(feat_vals[no])
+                if len(values) != len(unit_ids):
+                    break
+                flatten_vals = [item for sublist in values for item in sublist]
+                nspks_list = [sp for sp in nspikes.values()]
+                spikes_index = np.cumsum(nspks_list).astype("int64")
+                if ft in self.nwbfile.units:  # If property already exists, skip it
+                    warnings.warn(f"Feature {ft} already present in units table, skipping it")
+                    continue
+                set_dynamic_table_property(
+                    dynamic_table=self.nwbfile.units,
+                    row_ids=[int(k) for k in unit_ids],
+                    property_name=ft,
+                    values=flatten_vals,
+                    index=spikes_index,
+                )
+
+    def add_units_waveforms(self):
+        if self._get_unit_waveforms_templates(unit_id=0):
+            if len(self.nwbfile.units) == 0:
+                warnings.warn("create a units table before adding waveforms. Skipping operation")
+                return
+            units = self.nwbfile.units
+            waveform_metrics = {"mean": "mean", "std": "sd"}
+            for mode in waveform_metrics:
+                # construct wavforms for all units:
+                templates_all = []
+                for id in units.id.data:
+                    templates_all.append(self._get_unit_waveforms_templates(unit_id=id, mode=mode))
+                set_dynamic_table_property(
+                    dynamic_table=units,
+                    row_ids=units.id.data,
+                    property_name=f"waveform_{waveform_metrics[mode]}",
+                    values=templates_all,
+                )
+
+    @abstractmethod
+    def add_epochs(self):
+        pass
+
+
+def get_num_spikes(units_table, unit_id):
+    """Return the number of spikes for chosen unit."""
+    ids = np.array(units_table.id[:])
+    indexes = np.where(ids == unit_id)[0]
+    if not len(indexes):
+        raise ValueError(f"{unit_id} is an invalid unit_id. Valid ids: {ids}.")
+    index = indexes[0]
+    if index == 0:
+        return units_table["spike_times_index"].data[index]
+    else:
+        return units_table["spike_times_index"].data[index] - units_table["spike_times_index"].data[index - 1]
```

Comment:
should this go in utils?

Reply:
yes, we can move this to `common_writer_tools.py`